Drop qjsonwrapper dependency

nyalldawson opened the pull request:
Rework the oauth2 config to/from json conversion to avoid QMetaObject introspection, as this is fragile and prone to breakage without any compilation warning or errors. This allows us to remove the qjsonwrapper dependency, and drop an unwanted embedded library.

nyalldawson committed:
Macro should be HAVE_GUI, not WITH_GUI

nyalldawson committed:
Fix build

nyalldawson committed:
Fix clazy warning

nyalldawson committed:
Avoid some unnecessary use of qjsonwrapper

nyalldawson committed:
Directly set serialization values for oauth2 config

Don't use QMetaObject introspection for this, it's too fragile
and subject to warning-less breakage.

nyalldawson committed:
Drop unused qjsonwrapper external library

nyalldawson commented:
FYI @rouault

rouault commented on src/auth/oauth2/core/qgsauthoauth2config.cpp:
```diff
@@ -319,29 +320,43 @@ bool QgsAuthOAuth2Config::loadConfigTxt(
-      }
 
-      QJsonWrapper::qvariant2qobject( variantMap, this );
+      setApiKey( variantMap.value( QStringLiteral( "apiKey" ) ).toString() );
```
what happens if the variantMap lacks one of the item ? Wouldn't it be preferable not to call the set method then ?

rouault commented on src/auth/oauth2/core/qgsauthoauth2config.cpp:
```diff
@@ -319,29 +320,43 @@ bool QgsAuthOAuth2Config::loadConfigTxt(
+      setApiKey( variantMap.value( QStringLiteral( "apiKey" ) ).toString() );
+      setClientId( variantMap.value( QStringLiteral( "clientId" ) ).toString() );
+      setClientSecret( variantMap.value( QStringLiteral( "clientSecret" ) ).toString() );
+      setConfigType( static_cast<ConfigType>( variantMap.value( QStringLiteral( "configType" ) ).toInt() ) );
```
this is undefined behaviour if the integer value is out of range of the enumeration

nyalldawson replied:
Ok, addressed!

rouault commented on src/auth/oauth2/core/qgsauthoauth2config.cpp:
```diff
@@ -319,29 +320,43 @@ bool QgsAuthOAuth2Config::loadConfigTxt(
+      setClientSecret( variantMap.value( QStringLiteral( "clientSecret" ) ).toString() );
+      setConfigType( static_cast<ConfigType>( variantMap.value( QStringLiteral( "configType" ) ).toInt() ) );
+      setDescription( variantMap.value( QStringLiteral( "description" ) ).toString() );
+      setGrantFlow( static_cast<GrantFlow>( variantMap.value( QStringLiteral( "grantFlow" ) ).toInt() ) );
```
same here

rouault commented on src/auth/oauth2/core/qgsauthoauth2config.cpp:
```diff
@@ -319,29 +320,43 @@ bool QgsAuthOAuth2Config::loadConfigTxt(
+      setRedirectPort( variantMap.value( QStringLiteral( "redirectPort" ) ).toInt() );
+      setRedirectUrl( variantMap.value( QStringLiteral( "redirectUrl" ) ).toString() );
+      setRefreshTokenUrl( variantMap.value( QStringLiteral( "refreshTokenUrl" ) ).toString() );
+      setAccessMethod( static_cast<AccessMethod>( variantMap.value( QStringLiteral( "accessMethod" ) ).toInt() ) );
```
same here

nyalldawson committed:
Only set properties when present in json

nyalldawson committed:
Harden enum access

nyalldawson commented:
@rouault all addressed!

nyalldawson committed:
Fix clang-tidy warning

nyalldawson committed:
Remove useless test

This is just testing a very low level Qt method which has changed behavior in Qt 6
